add scorecard basic-check-spec-test and olm-suite-test

rocrisp opened the pull request.

bcrochet committed:
Use a temp dir for the container tarball

For the tarball extraction, we will use a temporary directory, and
clean it up when the tests are complete.

Signed-off-by: Brad P. Crochet <brad@redhat.com>

komish committed:
migrate check that confirms the image has under 40 layers

Signed-off-by: Jose R. Gonzalez <josegonzalez89@gmail.com>

komish committed:
add preflight log to the .gitignore

Signed-off-by: Jose R. Gonzalez <josegonzalez89@gmail.com>

komish committed:
add test containerfiles to help evaluate tool functionality

Signed-off-by: Jose R. Gonzalez <josegonzalez89@gmail.com>

samira-barouti committed:
Fix the error handling, add a parameter for the layer count

bcrochet committed:
Disable the local image path

The local image path as the target image was inadvertently enabled. This is
causing all tests to fail. Comment it again.

Signed-off-by: Brad P. Crochet <brad@redhat.com>

bcrochet committed:
Make the runs_as_nonroot test more robust

If there was any output on stderr, this check would usually fail. This
patch makes it a bit more robust by separating stdout and stderr, and
just relying on the stdout stream. It also uses TrimSpace instead of
TrimSuffix, as the suffix wasn't always CRLF, but sometime just LF,
and this would also cause the check to fail.

Signed-off-by: Brad P. Crochet <brad@redhat.com>

bcrochet committed:
Force use of podman when running operator-sdk bundle validate

We want to default the usage of podman (at least for now) when running
the bundle validate.

Signed-off-by: Brad P. Crochet <brad@redhat.com>

rocrisp committed:
add scorecard-basic-check-spec-test

rocrisp committed:
add scorecard olm suite test

samira-barouti commented on certification/internal/shell/scorecard_basic-check-spec-test.go:
```diff
@@ -0,0 +1,135 @@
+        return false, err
+    }
+
+    err = os.Chdir(bundledir)
```
can we use `--dir` flag of `oc image extract` to extract the image in the `bundledir` directory, instead of changing the working directory?

bcrochet replied:
I concur. Changing the workdir could be problematic.

rocrisp replied:
I removed change dir except for one place because operator-sdk requires to be in the parent's of the bundle location.

samira-barouti commented on certification/internal/shell/scorecard_basic-check-spec-test.go:
```diff
@@ -0,0 +1,135 @@
+    }
+    defer os.RemoveAll(tmpdir)
+
+    operatordir := tmpdir+"/test-operator-bundle"
```
Nit picking: if operatordir is not being reused, I suggest to merge the LoC 35 and 36

bcrochet replied:
And use filepath.Join

rocrisp replied:
Sounds good.

samira-barouti commented on certification/internal/shell/scorecard_basic-check-spec-test.go:
```diff
@@ -0,0 +1,135 @@
+        return false, err
+    }
+
+    fmt.Println("Running Scorecard test for ",bundle)
```
I suggest using logger.Debugf() instead of fmt.print

rocrisp replied:
Done.

samira-barouti commented on certification/internal/shell/scorecard_basic-check-spec-test.go:
```diff
@@ -0,0 +1,135 @@
+        return false, err
+    }
+    
+    return true, nil
```
Out of curiosity: is there any case that no error occurs and the validation fails? I do not see any `return false, nil` and was curious to understand if this won't happen at all

rocrisp replied:
This is what I found. 

Exit Status
The scorecard return code is 1 if any of the tests executed did not pass and 0 if all selected tests pass.
https://sdk.operatorframework.io/docs/advanced-topics/scorecard/scorecard/#exit-status.

samira-barouti commented on certification/internal/shell/scorecard_olm_suite.go:
```diff
@@ -0,0 +1,134 @@
+
+func (p *ScorecardOlmSuiteTest) Help() certification.HelpText {
+    return certification.HelpText{
+        Message:    "Operator-sdk scorecard OLM Test Suite. One more more test failed.",
```
typo?

rocrisp replied:
Can you be more specific?

samira-barouti replied:
`One more more`?

samira-barouti commented on certification/internal/shell/scorecard_olm_suite.go:
```diff
@@ -0,0 +1,134 @@
+package shell
```
The same comments as above for this Check :)

rocrisp replied:
Gotcha.

bcrochet commented on certification/internal/shell/scorecard_basic-check-spec-test.go:
```diff
@@ -0,0 +1,135 @@
+package shell
```
Let's be consistent in filenames. Use all underscores. And don't use '-test' or '_test', as that would normally signify a golang file that contains unit (or some other) test.

rocrisp replied:
I removed all references to test.

bcrochet commented on certification/internal/shell/scorecard_basic-check-spec-test.go:
```diff
@@ -0,0 +1,135 @@
+        return false, err
+    }
+    
+    _, err = exec.Command("chmod","-R","go+r","./").CombinedOutput()
```
For readability, please add a space after each comma.

rocrisp replied:
Done.

bcrochet commented on certification/internal/shell/scorecard_basic-check-spec-test.go:
```diff
@@ -0,0 +1,135 @@
+    fmt.Println("Running Scorecard test for ",bundle)
+    
+    stdouterr, err = exec.Command("operator-sdk", "scorecard",   
+                                  "--config",destinationfile,
```
Spaces after commas please.

rocrisp replied:
Done.

bcrochet commented on certification/internal/shell/scorecard_basic-check-spec-test.go:
```diff
@@ -0,0 +1,135 @@
+}
+
+func (p *ScorecardBasicCheckSpecTest) Name() string {
+    return "ScorecardBasicCheckSpecTest"
```
Remove 'Test'

rocrisp replied:
Ditto.

bcrochet commented on certification/internal/shell/scorecard_basic-check-spec-test.go:
```diff
@@ -0,0 +1,135 @@
+    "github.com/sirupsen/logrus"
+)
+
+type ScorecardBasicCheckSpecTest struct{}
```
Don't add 'Test'.

rocrisp replied:
no more test. :-)

bcrochet commented on certification/internal/shell/scorecard_olm_suite.go:
```diff
@@ -0,0 +1,134 @@
+    "github.com/sirupsen/logrus"
+)
+
+type ScorecardOlmSuiteTest struct{}
```
Remove the 'Test' part.

bcrochet commented on certification/internal/shell/scorecard_olm_suite.go:
```diff
@@ -0,0 +1,134 @@
+func (p *ScorecardOlmSuiteTest) Validate(bundle string, logger *logrus.Logger) (bool, error) {
+    
+    //Curentdir
+    homedir, err := os.Getwd()
```
Check for the error before you use homedir.

rocrisp replied:
Done.

bcrochet commented on certification/internal/shell/scorecard_olm_suite.go:
```diff
@@ -0,0 +1,134 @@
+    artifactsdir := homedir+"/artifacts"
+    test := "operator_bundle_scorecard_OlmSuiteTest.json"
+    
+    if err != nil {
```
Move this block up before line 19.

rocrisp replied:
Yep.

bcrochet commented on certification/internal/shell/scorecard_olm_suite.go:
```diff
@@ -0,0 +1,134 @@
+
+    fmt.Println("Running Scorecard test for ",bundle)
+    
+    scorecard_stdouterr, scorecard_err := exec.Command("operator-sdk", "scorecard",   
```
Spaces after commas.

rocrisp replied:
Done.

bcrochet commented on certification/internal/shell/scorecard_olm_suite.go:
```diff
@@ -0,0 +1,134 @@
+
+func (p *ScorecardOlmSuiteTest) Metadata() certification.Metadata {
+    return certification.Metadata{
+        Description:      "This test make sure that all CRs have a spec block.",
```
Rephrase to use the term 'check' instead of 'test' for consistency.

rocrisp replied:
Done.

bcrochet commented on certification/internal/shell/scorecard_basic-check-spec-test.go:
```diff
@@ -0,0 +1,135 @@
+
+func (p *ScorecardBasicCheckSpecTest) Metadata() certification.Metadata {
+    return certification.Metadata{
+        Description:      "This test make sure that all CRs have a spec block.",
```
Rephrase to use 'check' instead of 'test'

rocrisp replied:
Done.

samira-barouti commented on certification/internal/shell/scorecard_basic-check-spec-test.go:
```diff
@@ -0,0 +1,135 @@
+        return false, err
+    }
+
+    err = os.Chdir(tmpdir)
```
The same comment as above here; can we pass the `tempdir` to ReadFile instead of changing working dir?

rocrisp replied:
operator-sdk scorecard expects current dir so this is the only one  I kept for Chdir.

samira-barouti commented on certification/internal/shell/scorecard_basic-check-spec-test.go:
```diff
@@ -0,0 +1,135 @@
+        return false, err
+    }
+
+    err = os.Chdir(homedir)
```
Same comment here

rocrisp replied:
I removed it.

rocrisp committed:
fix PR suggestions

make changes per Pull Request

query json for scorecard result

add json parse logic

Change message

Correct the loop end

line things up

line things up

line things up

update/fix per PR

remove modules not needed

samira-barouti commented on certification/internal/shell/scorecard_basic-check-spec.go:
```diff
@@ -0,0 +1,123 @@
+
+type ScorecardBasicSpecCheck struct{}
+
+func (p *ScorecardBasicSpecCheck) Validate(bundle string, logger *logrus.Logger) (bool, error) {
```
Does `bundle` refer to `bundleImage`? It would be great to have a more descriptive name :)

rocrisp replied:
I corrected.

samira-barouti commented on certification/internal/shell/scorecard_basic-check-spec.go:
```diff
@@ -0,0 +1,123 @@
+        return false, err
+    }
+
+    scorecardresult := "operator_bundle_scorecard_BasicSpecCheck.json"
```
I am not sure if we follow any variable naming convention (like camelCase). IMO, to improve the code readability I will use scorecardResult instead. @bcrochet What do you think?

rocrisp replied:
I updated all to cameCase

samira-barouti commented on certification/internal/shell/scorecard_basic-check-spec.go:
```diff
@@ -0,0 +1,123 @@
+        return false, err
+    }
+
+    scorecardresult := "operator_bundle_scorecard_BasicSpecCheck.json"
```
This declaration could be in a `const` block.

rocrisp replied:
Done.

samira-barouti commented on certification/internal/shell/scorecard_olm_suilte.go:
```diff
@@ -0,0 +1,122 @@
+        return false, err
+    }
+
+	if scorecarderr != nil {
```
indentation is a bit off

rocrisp replied:
I removed the offending spaces.

samira-barouti commented on certification/internal/shell/scorecard_olm_suilte.go:
```diff
@@ -0,0 +1,122 @@
+        logger.Error("unable to execute operator-sdk scorecard: ", err)
+        return false, err
+    }
+    
```
I feel you should do a check on the output file to ensure all tests pass. You can search for `"state": "pass"` in the output file to return `true,nil` or `false,nil`

rocrisp replied:
I added guardrails against them.

samira-barouti commented on certification/internal/shell/scorecard_basic-check-spec.go:
```diff
@@ -0,0 +1,174 @@
+        var s string = v.(string)
+        logger.Debug(s)
+        if strings.Contains(s, "fail") {
+            foundTestFailed = true
```
I think you can have a `break` here as well, as soon as you detect a test failure.

rocrisp replied:
My thought with this is to flush out all the test run by scorecard hence the boolean variable for later to consume.
Example : 
time="2021-06-30T20:13:27Z" level=debug msg="output:olm-crds-have-resources"
time="2021-06-30T20:13:27Z" level=debug msg="output:fail"
time="2021-06-30T20:13:27Z" level=debug msg="output:olm-spec-descriptors"
time="2021-06-30T20:13:27Z" level=debug msg="output:fail"
time="2021-06-30T20:13:27Z" level=debug msg="output:olm-crds-have-validation"
time="2021-06-30T20:13:27Z" level=debug msg="output:pass"
time="2021-06-30T20:13:27Z" level=debug msg="output:olm-bundle-validation"
time="2021-06-30T20:13:27Z" level=debug msg="output:pass"
time="2021-06-30T20:13:27Z" level=debug msg="output:olm-status-descriptors"
time="2021-06-30T20:13:27Z" level=debug msg="output:fail"

samira-barouti commented on certification/internal/shell/scorecard_basic-check-spec.go:
```diff
@@ -0,0 +1,174 @@
+        return false, nil
+    }
+
+    if scorecarderr != nil {
```
I suggest you move this `if` block to line 93. Handling the erroneous condition must be done right after the call.

rocrisp replied:
So scorecard can failed because one or more test failed to run. The output contains information pertaining to which test failed to run. For now, I store the output in artifacts/ and it's information the developer can use to fix the problems.
With that said, if I exit at line 93, I will have no way of digesting this information.

samira-barouti commented on certification/internal/shell/scorecard_basic-check-spec.go:
```diff
@@ -0,0 +1,174 @@
+            break
+        }
+        if err, ok := v.(error); ok {
+            logger.Error(err)
```
can you please add a bit more context around this error? An error message will be helpful.

rocrisp replied:
Done.

samira-barouti commented on certification/internal/shell/scorecard_basic-check-spec.go:
```diff
@@ -0,0 +1,174 @@
+
+    logger.Debug("scorecard output")
+    
+    for {
```
To have a clean code, I suggest you have `return` statements inside the for loop, if it makes sense. Sth like:

 ```
for {
        v, ok := iter.Next()
        if !ok {
            break
        }
        if err, ok := v.(error); ok {
            logger.Error(err)
        }
        var s string = v.(string)
        logger.Debug(s)
        if strings.Contains(s, "fail") {
            return false, nil
      }
    }
return true, nil
```

rocrisp replied:
I can but this is a design intention to capture the tests in the output.
We can talk about this if you want to. I welcome suggestions of course.

samira-barouti replied:
Of course! Let's book sometimes tomorrow, Friday July 2nd to talk about rewriting this block.

samira-barouti commented on certification/internal/shell/scorecard_basic-check-spec.go:
```diff
@@ -0,0 +1,174 @@
+            logger.Error(err)
+        }
+        var s string = v.(string)
+        logger.Debug(s)
```
seems to be a leftover from the debugging :)

rocrisp replied:
I change it to logger.Info. instead

samira-barouti commented on certification/internal/shell/scorecard_basic-check-spec.go:
```diff
@@ -0,0 +1,174 @@
+        }
+    }
+
+    if foundTestFailed {
```
you can do `return !foundTestFailed,nil` and remove this block as well as line 153.

rocrisp replied:
I corrected it.

bcrochet commented on certification/internal/shell/scorecard_basic-check-spec.go:
```diff
@@ -0,0 +1,174 @@
+package shell
```
Filename should have _ not -.

rocrisp replied:
Corrected.

bcrochet commented on certification/internal/shell/scorecard_basic-check-spec.go:
```diff
@@ -0,0 +1,174 @@
+        if err, ok := v.(error); ok {
+            logger.Error(err)
+        }
+        var s string = v.(string)
```
Not a whole lot of reason to use the var syntax if you are assigning immediately. 

Just use s := v.(string)

rocrisp replied:
Ah! yes. I corrected.

bcrochet commented on certification/internal/shell/scorecard_olm_suilte.go:
```diff
@@ -0,0 +1,177 @@
+package shell
```
suite in the filename. Not suilte. :)

rocrisp replied:
Opps! Thanks.

bcrochet reviewed (commented):
A couple of comments. And eventually this needs to be collapsed into one commit.

rocrisp committed:
remove oc image extract because operator-sdk take an image

rocrisp committed:
rephrase

rocrisp committed:
add approve image registry check

rocrisp committed:
loggin

rocrisp committed:
remove operator_imagesource.go

rocrisp commented:
Too many changes so I will close this one and open a new PR
